6주차 과제 - 레스토랑 상세 페이지 구현하기

sejineeee opened the pull request:
## 내용
react-router-dom을 사용하기 위해 설치하였습니다.
아직 테스트 파일이 작성되지 않은 상태여서 route에 연결한 컴포넌트들은 푸시하지 않았습니다!
마저 작업하는대로 바로바로 올리고 질문이 있을 경우 같이 올려놓도록 하겠습니다!
이번주도 잘 부탁드립니다!!

sejineeee committed:
chore: react-router-dom 설치

sejineeee committed:
BrowserRouter 사용

App 컴포넌트를 BrowserRouter로 감싸서 App에서 선언하지 않고, index에서 선언

sejineeee committed:
header, routes, route 연결

Link를 이용해서 이동할 수 있게 구현
routes 하위에 route로 연결

daadaadaah reviewed (changes requested):
좋습니당!! 빠른 PR 보내기 잘하셨어요👏👏👏
다음 PR 기다리고 있을게여ㅎㅎ

sejineeee committed:
test: AboutPage 렌더링 테스트 구현

sejineeee committed:
feat: AboutPage 구현

sejineeee committed:
test: HomePage 렌더링 test 구현

sejineeee committed:
feat: HomePage 구현

sejineeee committed:
test: RestaurantsPage 렌더링 test 구현

sejineeee committed:
feat: RestaurantsPage 구현

기존에 App 컴포넌트에 있던 컴포넌트들을 RestaurantsPage로 분리

sejineeee committed:
test: MemoryRouter 사용, 렌더링시 헤더 텍스트 있는지 test

sejineeee committed:
restaurants Route 연결

sejineeee committed:
test: MemoryRouter 사용

sejineeee committed:
Link를 이용해서 해당 id로 이동하도록 구현

react-router-dom에 있는 Link를 이용해서 해당 id로 이동되도록 구현

sejineeee committed:
detail 테스트용 데이터생성

sejineeee committed:
restaurantDetail 컴포넌트 라우트 연결

sejineeee committed:
test: fetchRestaurantDetail api 호출 함수 test 구현

sejineeee committed:
fetchRestaurantDetail api 호출 함수 구현

sejineeee committed:
setRestaurantDetail action 추가

sejineeee committed:
loadRestaurantDetail, setRestaurantDetail 추가

sejineeee committed:
feat: RestaurantDetail 페이지 구현

useParams 이용, loadRestaurantDetail 함수에 id 전달
일단 데이터 들어오는지 확인을 위해 콘솔 출력

sejineeee commented:
## 내용
안녕하세요!!
일단 디테일 페이지를 라우트 연결하는 것까지는 완료했습니다.
그러고 디테일 페이지 데이터 api 호출 함수까지 구현하였습니다!
하지만 생각처럼 되지 않아서 ...ㅎㅎㅎ

처음에 state is not a function 이라는 에러가 뜨길래, state를 콘솔에 확인해봐야겠다 싶어서
useSelector 안에서 확인해봤는데... 😭 state는 확인이 되긴 하지만
현재 restaurantDetail에는 초기값([])만 들어있고, 데이터가 들어오질 않습니다ㅠㅠ
일단 PR을 남긴 후, 해결하기 위해 더 노력해보도록 하겠습니다...
<img width="742" alt="스크린샷 2022-11-16 오후 9 08 01" src="https://user-images.githubusercontent.com/86041335/202177415-789e3cda-09fc-472f-ae18-773ab7f43220.png">

sejineeee committed:
test: setRestaurantDetail test 구현

toEqual을 사용해서 만들어둔 테스트용 데이터와 같은지 테스트
하지만, 빈 배열이 담겨서 테스트 실패

sejineeee committed:
test: RestaurantDetail 테스트 구현

해당하는 텍스트가 있는지 테스트하였으나 실패
아무것도 렌더링되지 않는 것으로 테스트 확인됨

sejineeee committed:
데이터 받아지지 않는 오류 고침, 데이터 받아서 map을 이용하여 메뉴 구현

loadRestaurantDetail에 전달하는 인자 id를 id.id로 전달
if문을 이용하여 restaurantDetail이 빈 배열일 때, null 값 리턴
menuItems를 map을 돌리려고 하니 오류로 인해 optional chaining 메서드 사용

sejineeee commented on src/RestaurantDetail.jsx:
```diff
@@ -0,0 +1,35 @@
+      <p>{restaurantDetail.address}</p>
+      <ul>
+        {
+          restaurantDetail.menuItems?.map((item) => (<li key={item.id}>{item.name}</li>))
```
이 부분에 undefined로 인해서 옵셔널 체이닝을 이용해서 일단 문제를 해결하였습니다!
옵셔널 체이닝밖에 떠오르지 않아서 사용하였는데, 최선의 방법일지 잘 모르겠습니다ㅠㅠ

daadaadaah commented on src/RestaurantDetail.jsx:
```diff
@@ -0,0 +1,35 @@
+import { loadRestaurantDetail } from './actions';
+
+export default function RestaurantDetail() {
+  const id = useParams();
```
디스트러처링을 활용해보시는건 어떠신가여? 14번째줄이 그럼 깔끔해질 거 같은데
```js
  const { id } = useParams();
```

daadaadaah commented on src/RestaurantDetail.jsx:
```diff
@@ -0,0 +1,35 @@
+    restaurantDetail: state.restaurantDetail,
+  }));
+
+  if (restaurantDetail === []) {
```
length를 활용해봅시당!

daadaadaah commented on src/RestaurantDetail.test.jsx:
```diff
@@ -0,0 +1,28 @@
+import { useDispatch, useSelector } from 'react-redux';
+
+import RestaurantDetail from './RestaurantDetail';
+import RESTAURANTDETAIL from '../fixtures/restaurantDetail';
```
두 단어가 합쳐진 경우, `RESTAURANT_DETAIL` 이렇게 구분자로 구별해주면 좋아요!

daadaadaah reviewed (changes requested):
계속 진행해주세요! 남겨주신 질문은 해결 되신건가여?ㅎㅎ

sejineeee committed:
review: destructuring, 조건문 조건을 빈배열에서 length 값 0 확인으로 변경

sejineeee committed:
review: 테스트용 데이터 변수명이 두 단어가 합쳐진 경우 _ 언더스코어 추가

테스트용 데이터 변수명이 두 단어가 합쳐진 경우일 때
구분자(언더스코어)를 이용하여 구별할 수 있도록 수정

sejineeee committed:
setRestaurantDetail action 테스트 수정

reducer에 setRestaurants로 전달한 액션을 setRestaurantDetail로 수정
RESTAURANTDETAIL에 구분자 추가하여 구별

sejineeee commented:
## 내용
리뷰를 토대로 수정하였습니다!!!
reducer에 action에 테스트 잘못 들어온 부분은 제가 액션을 잘못 줬기 때문에 생긴 문제로 수정하였습니다!
아직 RestaurantDetail 부분에는 왜 아무것도 렌더링되지 않는 것일까요...?
그리고 제가 질문을 같이 올렸었는데... 제가 아직 깃허브 사용이 익숙하지 않다 보니, 질문이 Pending 상태였더라고요! ㅎㅎㅎ
그래서 늦게나마, 다시 질문 올렸습니다(옵셔널 체이닝 질문입니다!)
감사합니다!!

sejineeee committed:
test: NotFoundPage 렌더링 test

sejineeee committed:
feat: NotFoundPage 구현

404 Not Found 문구가 보이는 NotFoundPage 구현

sejineeee committed:
test: 레스토랑 상세정보가 있을 경우, 없을 경우 나누어 테스트

나누어 테스트 하였으나, 상세 정보가 있을 경우 렌더링되지 않아 테스트 실패

sejineeee committed:
review: 테스트용 데이터 구분자 넣어서 수정

테스트용 데이터 RESTAURANTDETAIL에 구분자 넣어서 RESTAURANT_DETAIL로 수정

sejineeee committed:
test: api mock에 fetchRestaurantDetail 추가

sejineeee committed:
historyApiFallback 추가

공식문서를 참고하여 historyApiFallback 추가

daadaadaah reviewed (changes requested):
잘하셨습니다!ㅎㅎ테스트 코드 작성하시는 것도 이제 제법 익숙해주신 것 같아 좋네영!👍

## 질문에 대한 답변
1. 현재 `restaurantDetail`를 배열로 관리하시는데, 컴포넌트에서는 객체로 취급하고 있어요!ㅎㅎ또한, fixture 데이터에서 값이랑 테스트코드에서 확인하고자 하는 값이랑 일치하는지 확인해보셔야 할 것 같아요!ㅎㅎ

sejineeee committed:
review: 테스트용 데이터 배열에서 객체로 수정

배열 안에 객체가 담긴 구조에서, 배열을 제외하고 객체 구조로 수정

sejineeee committed:
review: 테스트용 데이터의 구조로 인한 오류 해결

테스트용 데이터의 구조로 인한 오류 해결
주석 제거
테스트에서 확인하는 값과 테스트용 데이터의 값 동일하게 수정

sejineeee committed:
구조분해 할당 적용하여 수정

restaurantDetail 변수를 구조분해 할당하여 name, address, menuItems로 사용

sejineeee commented:
## 내용
안녕하세요! 컴포넌트에서는 객체로 취급하고, fixture 데이터에서는 배열로 만들어서 구조가 달라서 그랬던 거였다니...!
어제 하루종일 봤는데 이런 간단한 문제였다니...충격이에요! ㅎㅎㅎㅎ
restaurantDetail에 배열에 객체가 담긴 구조라고 저는 생각했는데, 콘솔로 확인해보니 객체만 담긴 것을 확인했습니다!

그러면 reducer 파일에 초기 state 설정 값에 제가 빈배열을 할당했는데, 이렇게 `null`을 할당하는 편이 더 나을까요?
```javascript
// reducer.js

const initialState = {
  restaurantDetail : null
}
```
그리고 제가 `RestaurantDetail.jsx` 파일에서

```javascript
 if (restaurantDetail.length === 0) {
    return null;
  }
```
이렇게 코드를 구현했었는데, 이 부분도 이렇게 작성해도 괜찮은 것일까요..?
뭔가 배열 안에 객체가 담긴 구조가 아닌데 (처음에는 배열 안에 객체가 담긴 구조라고 생각해서 이렇게 처리를 했었습니다!)
이렇게 작성해도 괜찮은지 궁금합니다!

만약, 초기값을 null로 할당한다면 이렇게 바꿔야 되는 것일까요?
```javascript
if (restaurantDetail === null) {
   return null;
}
```

sejineeee committed:
NotFoundPage 라우트 연결

sejineeee committed:
test: loadRestaurantDetail 액션 테스트 구현

sejineeee committed:
test: loadRestaurantDetail 테스트 오류 수정

sejineeee committed:
test: 레스토랑 상세 정보가 있을 경우의 테스트 이름 변경

daadaadaah commented on src/RestaurantDetail.jsx:
```diff
@@ -0,0 +1,37 @@
+
+  const { name, address, menuItems } = restaurantDetail;
+
+  if (restaurantDetail.length === 0) {
+    return null;
+  }
```
이제 배열이 아니니, 질문주신대로 아래처럼 표현하거나
```js
if (restaurantDetail === null) {
   return null;
}
```
- 아래처럼 해주셔도 괜찮을 것 같아요! 추가로, 없을 때, null 보다는 사용자에게 좀더 친화적인 UI를 넣어주시는 것도 좋을 것 같아요!
```js
if (!restaurantDetail) {
   return <p>loaing...</p>;
}
```

daadaadaah reviewed (changes requested):
> e2e 테스트에서 이렇게 나오는데 다른 조건에 대한 테스트가 더 필요한 걸까요? useParams에 대한 조건도 테스트를 해줘야 되는 건가요?
- `http://localhost:8080/restaurants/1` 링크로 입력했을 떄에도 화면이 보여야 하는데, 안보여서 실패하는 것 같아요! 하루 시간 더 있으니, 혼자 스스로 해결해보시는 걸 추천드려요! 다른 분들도 동일한 문제를 겪었는데, 어떻게 해결하셨는지 살펴봐도 좋구요!ㅎㅎ

sejineeee committed:
main.js 경로 변경

index.html 파일에서 script src 주소 main.js에서 /main.js로 변경

sejineeee committed:
reveiw: restaurantDetail initialState를 빈 배열에서 null로 수정

sejineeee committed:
test: restaurantDetail initialState를 빈 배열에서 null로 변경

sejineeee committed:
test: restaurantDetail 값 null로 변경, 테스트 수정

아무것도 렌더링 되지 않는다에서 로딩 메시지 렌더링 테스트로 변경

sejineeee committed:
review: restaurantDetail의 초기 상태 변경으로 인해서 조건문 조건 변경

restaurantDetail의 초기 상태 null로 변경하여 조건문의 조건 변경
return문에서 null을 로딩문구를 보여주도록 변경

sejineeee commented:
## 내용
안녕하세요!! 에러 메시지를 구글링 해보니, `index.html`의 스크립트 경로를 `main.js`에서 `/main.js`로 바꾸면 되는 것을 찾을 수 있었습니다!!! (저는 처음에 url에 저주가 걸린 줄 알았어요ㅎㅎㅎ 어떤 것을 해도 /1만 붙으면 에러가 뜨길래 푸하하)
일단 구글링을 통해서 해결은 하였으나, 왜 경로 설정으로 인해서 문제가 해결이 된 건지 이해가 잘 되지 않습니다!
혹시 이러한 이유에 대해서 알고 싶다면 어떤 키워드를 검색하면 제가 이해하기에 도움이 될까요...?
알려주신다면 이러한 이유에 대해서 공부해보도록 하겠습니다!! 🤓

daadaadaah reviewed (approved):
sejineeee님이라면 혼자 해결하실 수 있을 것 같았어요!👍 잘하셨어요~

- `절대경로`와 `상대경로`의 차이에 대해 공부해보시면 될거예요!
- 추가로, `main.js`에 대한 문제는 `index.html`의 `head`에
`<base href="/">`를 추가해서 해결할 수도 있습니다.
- 참고 : https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base
